std/typecons.d

```diff
@@ -1184,7 +1184,7 @@ if (distinctFieldNames!(Specs))
             Returns:
                 A `size_t` representing the hash of this `Tuple`.
          */
-        size_t toHash() const nothrow @safe
+        size_t toHash()() const nothrow @safe
```

Comment:
```suggestion
        size_t toHash() const nothrow @safe
```

And below replace
```d
static if (__traits(compiles, h = .hashOf(field[i])))
    const k = .hashOf(field[i]);
else
    // Workaround for when .hashOf is not both @safe and nothrow.
    // BUG: Improperly casts away `shared`!
    const k = typeid(T).getHash((() @trusted => cast(const void*) &field[i])());
```

with something like

```d
static if (__traits(compiles, h = .hashOf(field[i])))
    const k = .hashOf(field[i]);
else static if (is(T : shared U, U) && __traits(compiles, h = .hashOf((() @trusted => cast(U) field[i])())))
    // BUG: Improperly casts away `shared`!
    const k = .hashOf((() @trusted => cast(U) field[i])());
else
    // Workaround for when .hashOf is not both @safe and nothrow.
    // BUG: Improperly casts away `shared`!
    const k = typeid(T).getHash((() @trusted => cast(const void*) &field[i])());
```

This change is reasonable because we also have to cast away shared to call `typeid(T).getHash`. So if removing shared would also let us call `.hashOf` there is no reason to involve `typeid`. This is enough to make the current tests compile and pass both with and without -betterC.